src/EditorFeatures/Core/Implementation/CallHierarchy/Finders/AbstractCallFinder.cs

```diff
@@ -48,51 +49,59 @@ public void CancelSearch()
-
             var asyncToken = _asyncListener.BeginAsyncOperation(this.GetType().Name + ".Search");
 
-            // Follow the search task with task that lets the callback know we're done and which
-            // marks the async operation as complete.  Note that we pass CancellationToken.None
-            // here.  That's intentional.  This operation is *not* cancellable.
+            // NOTE: This task has CancellationToken.None specified, since it must complete no matter what
+            // so the callback is appropriately notified that the search has terminated.
+            Task.Run(async () =>
+            {
+                // The error message to show if we had an error. null will mean we succeeded.
+                string completionErrorMessage = null;
+                try
+                {
+                    await SearchAsync(workspace, searchScope, callback, _cancellationSource.Token).ConfigureAwait(false);
+
+                }
+                catch (OperationCanceledException)
+                {
+                    completionErrorMessage = EditorFeaturesResources.Canceled;
+                }
+                catch (Exception e)
+                {
+                    completionErrorMessage = e.Message;
+                }
+                finally
+                {
+                    if (completionErrorMessage != null)
+                    {
+                        callback.SearchFailed(completionErrorMessage);
+                    }
+                    else
+                    {
+                        callback.SearchSucceeded();
+                    }
+                }
+            }, CancellationToken.None).CompletesAsyncOperation(asyncToken);
+        }
 
-            var workspace = _project.Solution.Workspace;
-            var currentProject = workspace.CurrentSolution.GetProject(_project.Id);
-            var compilation = await currentProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var resolution = _symbol.Resolve(compilation, cancellationToken: cancellationToken);
+        private async Task SearchAsync(Workspace workspace, CallHierarchySearchScope scope, ICallHierarchySearchCallback callback, CancellationToken cancellationToken)
+        {
+            var project = workspace.CurrentSolution.GetProject(_projectId);
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var resolution = _symbolKey.Resolve(compilation, cancellationToken: cancellationToken);
 
-            var documents = this.Documents ?? IncludeDocuments(scope, currentProject);
+            var documents = this.Documents ?? IncludeDocuments(scope, project);
 
             var currentSymbol = resolution.Symbol;
 
             if (currentSymbol == null)
             {
-                return;
+                throw new Exception(string.Format(WorkspacesResources.The_symbol_0_cannot_be_located_within_the_current_solution, SymbolName));
```

Comment:
ick.  can this be done without exceptions?  Seems like a non-exceptional case.

Reply:
Trying to avoid multiple exit paths. Given I'm fixing multiple bugs caused by an innocent "return".